api/openapi_server/repositories/service_provider_repository.py

```diff
@@ -20,121 +22,83 @@ def __init__(self, db_engine=None):
             db_engine = db.DataAccessLayer.get_engine()
         self.db_engine = db_engine
 
-    def create_service_provider(self, provider):
-        """Create a housing program service provider
-
-        :param provider: provider to create. 
-        :type provider: ServiceProvider
-
-        :rtype: ServiceProviderWithId
+    def create_service_provider(self, provider_name: str) -> Optional[HousingProgramServiceProvider]:
         """
-        session = self._get_session()
-        row = self._generate_row(provider)
-
-        session.add(row)
-        session.commit()
-        provider["id"] = row.id
+        Create a housing program service provider, if it
+        is not already in the database. 
 
-        session.close()
-        return ServiceProviderWithId.from_dict(provider)
+        Return the newly created service provider. Return None
+        if the service provider already exists.
+        """
+        with DataAccessLayer.session() as session:
+            existingProvider = session.query(HousingProgramServiceProvider).filter(
+                HousingProgramServiceProvider.provider_name == provider_name
+            ).one_or_none()
+
+            if existingProvider is None:
+                newProvider = HousingProgramServiceProvider(
+                    provider_name = provider_name
+                )
+                session.add(newProvider)
+                session.commit()
+                session.flush()
+                # primary keys are lazy loaded by default
+                _ = newProvider.id
```

Comment:
Because I read [Autogenerated primary key attributes](https://docs.sqlalchemy.org/en/14/tutorial/orm_data_manipulation.html#autogenerated-primary-key-attributes), I thought the primary key is already available by default.

Also, the sentence above that section says changes are flushed out when `commmit` is called.

Reply:
primary key is already available, but only in the database. You need to refresh the python object before id field is populated. See my explanation above for more detail

A better way of refreshing was added in [1d8a319](https://github.com/hackforla/HomeUniteUs/pull/575/commits/1d8a319d4d549ac1424390e5c468d8689aba3f97)